### PR TITLE
merge qkv matrices

### PR DESCRIPTION
**Pull request by tnguyen21**
```
tokens per iteration: 65,536
breakdown: config.gradient_accumulation_steps=1 * ddp_world_size=4 * config.batch_size=16 * config.max_seq_len=1024
total params: 6,682,124,288 parameters
  Muon (2D weights): 128 tensors, 4,865,392,640 parameters
  AdamW (others): 98 tensors, 1,816,731,648 parameters
Muon lr: 0.05, AdamW lr: 0.001
using fused AdamW: True
Python 3.11.10 (main, Dec  3 2024, 02:25:00) [GCC 12.2.0]
PyTorch 2.10.0.dev20251003+cu128 (CUDA 12.8)
Sun Oct  5 19:44:53 2025
+-----------------------------------------------------------------------------------------+
| NVIDIA-SMI 575.57.08              Driver Version: 575.57.08      CUDA Version: 12.9     |
|-----------------------------------------+------------------------+----------------------+
| GPU  Name                 Persistence-M | Bus-Id          Disp.A | Volatile Uncorr. ECC |
| Fan  Temp   Perf          Pwr:Usage/Cap |           Memory-Usage | GPU-Util  Compute M. |
|                                         |                        |               MIG M. |
|=========================================+========================+======================|
|   0  NVIDIA H100 80GB HBM3          On  |   00000000:0A:00.0 Off |                    0 |
| N/A   39C    P0            121W /  700W |   26661MiB /  81559MiB |      0%      Default |
|                                         |                        |             Disabled |
+-----------------------------------------+------------------------+----------------------+
|   1  NVIDIA H100 80GB HBM3          On  |   00000000:0B:00.0 Off |                    0 |
| N/A   37C    P0            117W /  700W |   25643MiB /  81559MiB |     32%      Default |
|                                         |                        |             Disabled |
+-----------------------------------------+------------------------+----------------------+
|   2  NVIDIA H100 80GB HBM3          On  |   00000000:84:00.0 Off |                    0 |
| N/A   39C    P0            118W /  700W |   26675MiB /  81559MiB |      0%      Default |
|                                         |                        |             Disabled |
+-----------------------------------------+------------------------+----------------------+
|   3  NVIDIA H100 80GB HBM3          On  |   00000000:85:00.0 Off |                    0 |
| N/A   36C    P0            117W /  700W |   27225MiB /  81559MiB |      0%      Default |
|                                         |                        |             Disabled |
+-----------------------------------------+------------------------+----------------------+

+-----------------------------------------------------------------------------------------+
| Processes:                                                                              |
|  GPU   GI   CI              PID   Type   Process name                        GPU Memory |
|        ID   ID                                                               Usage      |
|=========================================================================================|
|    0   N/A  N/A               1      C   /bin/dumb-init                        26652MiB |
|    1   N/A  N/A               1      C   /bin/dumb-init                        26150MiB |
|    2   N/A  N/A               1      C   /bin/dumb-init                        26722MiB |
|    3   N/A  N/A               1      C   /bin/dumb-init                        27216MiB |
+-----------------------------------------------------------------------------------------+

step 0: train loss 11.6675, val loss 11.6613
0 | loss 11.6875 | lr 1.000000e-03 | 94321.78ms
10 | loss 9.1875 | lr 1.000000e-03 | 3479.80ms
20 | loss 9.5625 | lr 1.000000e-03 | 3498.97ms
30 | loss 8.3750 | lr 1.000000e-03 | 3511.68ms
40 | loss 7.7188 | lr 1.000000e-03 | 3514.13ms
50 | loss 7.4375 | lr 1.000000e-03 | 3523.38ms
60 | loss 7.0000 | lr 9.118812e-04 | 3529.13ms
70 | loss 7.0625 | lr 7.138614e-04 | 3527.50ms
80 | loss 5.9375 | lr 5.158416e-04 | 3519.55ms
90 | loss 6.7500 | lr 3.178218e-04 | 3519.44ms
step 100: train loss 6.4956, val loss 6.4706
100 | loss 6.0625 | lr 1.198020e-04 | 90926.85ms
Return code: 0
```

- improves step time by ~300 ms
- does not improve rate of convergence. not sure why, it should be "equivalent" to if the matrices were separate. likely due to uniform init?

### CODE REVIEW

**Comment by tnguyen21** [comment]
```
tokens per iteration: 65,536
breakdown: config.gradient_accumulation_steps=1 * ddp_world_size=4 * config.batch_size=16 * config.max_seq_len=1024
total params: 6,682,124,288 parameters
  Muon (2D weights): 160 tensors, 6,476,005,376 parameters
  AdamW (others): 66 tensors, 206,118,912 parameters
Muon lr: 0.05, AdamW lr: 0.001
using fused AdamW: True
Python 3.11.10 (main, Dec  3 2024, 02:25:00) [GCC 12.2.0]
PyTorch 2.10.0.dev20251003+cu128 (CUDA 12.8)
Sun Oct  5 20:15:05 2025
+-----------------------------------------------------------------------------------------+
| NVIDIA-SMI 575.57.08              Driver Version: 575.57.08      CUDA Version: 12.9     |
|-----------------------------------------+------------------------+----------------------+
| GPU  Name                 Persistence-M | Bus-Id          Disp.A | Volatile Uncorr. ECC |
| Fan  Temp   Perf          Pwr:Usage/Cap |           Memory-Usage | GPU-Util  Compute M. |
|                                         |                        |               MIG M. |
|=========================================+========================+======================|
|   0  NVIDIA H100 80GB HBM3          On  |   00000000:05:00.0 Off |                    0 |
| N/A   32C    P0            121W /  700W |   26645MiB /  81559MiB |      5%      Default |
|                                         |                        |             Disabled |
+-----------------------------------------+------------------------+----------------------+
|   1  NVIDIA H100 80GB HBM3          On  |   00000000:0B:00.0 Off |                    0 |
| N/A   35C    P0            123W /  700W |   27209MiB /  81559MiB |      0%      Default |
|                                         |                        |             Disabled |
+-----------------------------------------+------------------------+----------------------+
|   2  NVIDIA H100 80GB HBM3          On  |   00000000:0C:00.0 Off |                    0 |
| N/A   32C    P0            123W /  700W |   26645MiB /  81559MiB |      0%      Default |
|                                         |                        |             Disabled |
+-----------------------------------------+------------------------+----------------------+
|   3  NVIDIA H100 80GB HBM3          On  |   00000000:84:00.0 Off |                    0 |
| N/A   42C    P0            121W /  700W |   26159MiB /  81559MiB |     27%      Default |
|                                         |                        |             Disabled |
+-----------------------------------------+------------------------+----------------------+

+-----------------------------------------------------------------------------------------+
| Processes:                                                                              |
|  GPU   GI   CI              PID   Type   Process name                        GPU Memory |
|        ID   ID                                                               Usage      |
|=========================================================================================|
|    0   N/A  N/A               1      C   /bin/dumb-init                        26636MiB |
|    1   N/A  N/A               1      C   /bin/dumb-init                        27200MiB |
|    2   N/A  N/A               1      C   /bin/dumb-init                        26698MiB |
|    3   N/A  N/A               1      C   /bin/dumb-init                        26150MiB |
+-----------------------------------------------------------------------------------------+

step 0: train loss 11.6250, val loss 11.6275
0 | loss 11.6250 | lr 1.000000e-03 | 95337.09ms
10 | loss 9.8125 | lr 1.000000e-03 | 3708.98ms
20 | loss 8.6250 | lr 1.000000e-03 | 3642.29ms
30 | loss 7.2188 | lr 1.000000e-03 | 3639.89ms
40 | loss 7.1562 | lr 1.000000e-03 | 3650.02ms
50 | loss 6.9375 | lr 1.000000e-03 | 3638.96ms
60 | loss 6.6250 | lr 9.118812e-04 | 3650.84ms
70 | loss 6.7500 | lr 7.138614e-04 | 3641.77ms
80 | loss 5.5625 | lr 5.158416e-04 | 3649.22ms
90 | loss 6.4062 | lr 3.178218e-04 | 3646.99ms
step 100: train loss 6.1600, val loss 6.1350
100 | loss 5.7188 | lr 1.198020e-04 | 89665.04ms
```

this is if we change:

`self.qkv = nn.Linear(args.dim, 3 * hdim, bias=False)`

and in the fwd

`qkv = self.qkv(x)`

perhaps we can get the best of both worlds by using nn.Parameter with the init values that's default for linear layers?

**Comment by tnguyen21** [comment]
```
tokens per iteration: 65,536
breakdown: config.gradient_accumulation_steps=1 * ddp_world_size=4 * config.batch_size=16 * config.max_seq_len=1024
total params: 6,682,124,288 parameters
  Muon (2D weights): 128 tensors, 4,865,392,640 parameters
  AdamW (others): 98 tensors, 1,816,731,648 parameters
Muon lr: 0.05, AdamW lr: 0.001
using fused AdamW: True
Python 3.11.10 (main, Dec  3 2024, 02:25:00) [GCC 12.2.0]
PyTorch 2.10.0.dev20251003+cu128 (CUDA 12.8)
Sun Oct  5 20:44:47 2025
+-----------------------------------------------------------------------------------------+
| NVIDIA-SMI 575.57.08              Driver Version: 575.57.08      CUDA Version: 12.9     |
|-----------------------------------------+------------------------+----------------------+
| GPU  Name                 Persistence-M | Bus-Id          Disp.A | Volatile Uncorr. ECC |
| Fan  Temp   Perf          Pwr:Usage/Cap |           Memory-Usage | GPU-Util  Compute M. |
|                                         |                        |               MIG M. |
|=========================================+========================+======================|
|   0  NVIDIA H100 80GB HBM3          On  |   00000000:0C:00.0 Off |                    0 |
| N/A   33C    P0            123W /  700W |   26661MiB /  81559MiB |      3%      Default |
|                                         |                        |             Disabled |
+-----------------------------------------+------------------------+----------------------+
|   1  NVIDIA H100 80GB HBM3          On  |   00000000:84:00.0 Off |                    0 |
| N/A   36C    P0            117W /  700W |   26267MiB /  81559MiB |      0%      Default |
|                                         |                        |             Disabled |
+-----------------------------------------+------------------------+----------------------+
|   2  NVIDIA H100 80GB HBM3          On  |   00000000:85:00.0 Off |                    0 |
| N/A   33C    P0            118W /  700W |   27225MiB /  81559MiB |      5%      Default |
|                                         |                        |             Disabled |
+-----------------------------------------+------------------------+----------------------+
|   3  NVIDIA H100 80GB HBM3          On  |   00000000:8B:00.0 Off |                    0 |
| N/A   40C    P0            122W /  700W |   27225MiB /  81559MiB |      7%      Default |
|                                         |                        |             Disabled |
+-----------------------------------------+------------------------+----------------------+

+-----------------------------------------------------------------------------------------+
| Processes:                                                                              |
|  GPU   GI   CI              PID   Type   Process name                        GPU Memory |
|        ID   ID                                                               Usage      |
|=========================================================================================|
|    0   N/A  N/A               1      C   /bin/dumb-init                        26652MiB |
|    1   N/A  N/A               1      C   /bin/dumb-init                        26258MiB |
|    2   N/A  N/A               1      C   /bin/dumb-init                        27216MiB |
|    3   N/A  N/A               1      C   /bin/dumb-init                        27216MiB |
+-----------------------------------------------------------------------------------------+

step 0: train loss 11.6663, val loss 11.6625
0 | loss 11.6875 | lr 1.000000e-03 | 93902.98ms
10 | loss 9.7500 | lr 1.000000e-03 | 3455.56ms
20 | loss 9.5000 | lr 1.000000e-03 | 3481.03ms
30 | loss 7.9062 | lr 1.000000e-03 | 3496.33ms
40 | loss 7.7500 | lr 1.000000e-03 | 3504.67ms
50 | loss 7.2812 | lr 1.000000e-03 | 3505.70ms
60 | loss 6.9375 | lr 9.118812e-04 | 3510.31ms
70 | loss 7.0312 | lr 7.138614e-04 | 3503.73ms
80 | loss 5.9062 | lr 5.158416e-04 | 3512.46ms
90 | loss 6.6875 | lr 3.178218e-04 | 3504.79ms
step 100: train loss 6.4619, val loss 6.4363
100 | loss 6.0312 | lr 1.198020e-04 | 90476.64ms
Return code: 0
```

trying to use nn.Parameter with a similar init distribution as nn.Linear still does far worse than nn.Linear. must be something else going on there that might be worth need to diging into
